### DISCUSSION TITLE
[Flex] Private recipe and copy-from-recipe

### DISCUSSION THREAD

**Discussion by lugosium** [symfony/symfony#44718] [Q&A] [answered]
Hello,

I wanted to create my first flex private recipe and followed [documentation](https://symfony.com/doc/current/setup/flex_private_recipes.html) but i struggle with the `copy-from-recipe` feature.

It does not copy my file and i have no errors (installed with -vvv nothing in logs).

The recipe is correctly executed for `env`, `files`, `copy-from-package` but i cannot figure out why it does not work with `copy-from-recipe`.

I have:

`lugosium.inpulsar-bundle.1.0.json`:
```
{
  "manifests": {
    "lugosium/inpulsar-bundle": {
      "manifest": {
        "copy-from-recipe": {
          "config/": "%CONFIG_DIR%"
        },
      },
      "ref": "f89cc2fe71911480795e6a9be245801cc76d5c80"
    }
  }
}
```

`index.json`:

```
{
  "recipes": {
    "lugosium/inpulsar-bundle": [
      "1.0"
    ]
  },
  "branch": "master",
  "is_contrib": true,
  "_links": {
    "repository": "github.com/my-repo/my-recipes",
    "origin_template": "{package}:{version}@github.com/my-repo/my-recipes:master",
    "recipe_template": "https://api.github.com/repos/my-repo/my-recipes/contents/{package_dotted}.{version}.json"
  }
}
```

So i try with this structure:

```
- config
|_ file-to-copy.json
- index.json
- lugosium.inpulsar-bundle.1.0.json
```

Or

```
- lugosium
|_ inpulsar-bundle
  |_ 1.0
    |_ config
      |_ file-to-copy.json
- index.json
- lugosium.inpulsar-bundle.1.0.json
```

When installing i have:

```
- Configuring lugosium/inpulsar-bundle (>=1.0): From github.com/my-repo/my-recipes:master
    Copying files from recipe
```

But it end up like it has no files to copy. 

Can someone explains? Thank you.

**Comment by waxcage** [comment]
Have you resolved your issue? I am having the same problem and tried all possibilities and still I am not able to use copy-from-recipe.

**Reply by lugosium** [reply]
Not resolved. I don’t know why.

**Comment by romainguerrero** [comment, marked as answer]
Hi,

Following what was explained in the blog post https://symfony.com/blog/symfony-flex-is-going-serverless, it is really simple (when we understand how it works 😅) !

You have to use (almost) 2 branches in your recipes repo : `main` and `flex/main` like in the [official](https://github.com/symfony/recipes) or the [contrib](https://github.com/symfony/recipes-contrib) repos :
- In the `main` branch you can save your recipes, arranged by folders where you can save your files you want to copy
- The `flex/main` branch will contains only the json files like you have actually

To help you with those two branches, you can use the `flex-update` Github Action which will collect all informations you push on the `main` branch to push automatically what is needed in the `flex/main` branch.

You juste have to create the `flex/main` branch and put a file `.github/workflows/flex-update.yml` on your `main` branch containing the following content :
```yaml
name: Update Flex endpoint

on:
    push:
        branches:
            - main

defaults:
    run:
        shell: bash

jobs:
    call-flex-update:
        uses: symfony/recipes/.github/workflows/callable-flex-update.yml@main
        with:
            contrib: true
```

Then modify your recipe to have the architecture `lugosium/inpulsar-bundle/1.0` with a file `manifest.json` containing only :
```json
{
    "copy-from-recipe": {
        "config/": "%CONFIG_DIR%"
    }
}
```
You can then create your config folder next to the `manifest.json` file and when you will push, the action will automatically update the `flex/main` files accordingly 🎉 

Thanks to [yivi](https://stackoverflow.com/users/1426539/yivi) and [Spomky-Labs](https://stackoverflow.com/users/2157818/spomky-labs) for their answers [here](https://stackoverflow.com/a/71761865) and [here](https://stackoverflow.com/a/71693141/1426539) thanks to whom I was able to better understand how it works

**Reply by lugosium** [reply]
Wow! Thanks a lot Romain for your answer 👍 🥳
